File: leetcode/solutions/0088.MergeSortedArray/merge.go
package solution

func merge(nums1 []int, m int, nums2 []int, n int) {
	i := 0

	for j, num := range nums2 {
		for ; i < m; i++ {
			if num < nums1[i] {
				break
			}
		}

		if i == m {
			copy(nums1[m:], nums2[j:])
			return
		}

		copy(nums1[i+1:], nums1[i:])
		nums1[i] = num
		i++
		m++
	}
}

File: leetcode/solutions/0088.MergeSortedArray/merge.ts
export default function merge(
  nums1: number[],
  m: number,
  nums2: number[],
  n: number
): void {
  let i = m - 1;
  let j = n - 1;
  let k = n + m - 1;

  for (; i >= 0 && j >= 0; k--) {
    if (nums1[i] > nums2[j]) {
      nums1[k] = nums1[i];
      i--;
    } else {
      nums1[k] = nums2[j];
      j--;
    }
  }

  while (j >= 0) {
    nums1[k--] = nums2[j--];
  }
}
